python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,346 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import sys
+from typing import Iterable, Tuple
+
+if sys.version_info >= (3, 8):  # pragma: no cover
+    from functools import singledispatchmethod
+    from typing import Protocol
+else:  # pragma: no cover
+    from typing_extensions import Protocol  # type: ignore
+    from singledispatch import singledispatchmethod  # type: ignore
+
+from iceberg.types import IcebergType, ListType, MapType, NestedField, StructType
+
+
+class SchemaVisitor(Protocol):
+    def visit(self, node) -> None:  # pragma: no cover
+        ...
+
+
+class Schema:
+    """Schema of a table
+
+    Example:
+        >>> from iceberg.table.schema import Schema
+        >>> from iceberg.types import BooleanType, IntegerType, NestedField, StringType
+        >>> fields = [ \
+            NestedField(field_id=1, name="foo", field_type=StringType(), is_optional=False), \
+            NestedField(field_id=2, name="bar", field_type=IntegerType(), is_optional=True), \
+            NestedField(field_id=3, name="baz", field_type=BooleanType(), is_optional=False), \
+        ]
+        >>> table_schema = Schema(fields=fields, schema_id=1, aliases={"qux": 3})
+        >>> print(table_schema)
+        1: name=foo, type=string, required=True
+        2: name=bar, type=int, required=False
+        3: name=baz, type=boolean, required=True
+    """
+
+    def __init__(self, fields: Iterable[NestedField], schema_id: int, aliases: dict = {}):
+        self._struct = StructType(*fields)
+        self._schema_id = schema_id
+        self._aliases = aliases
+
+        index_by_id_visitor = IndexById()
+        index_by_id_visitor.visit(self._struct)
+        self._index_by_id = index_by_id_visitor.result
+
+        index_by_name_visitor = IndexByName()
+        index_by_name_visitor.visit(self._struct)
+        self._index_by_name = index_by_name_visitor.result
+
+    def __len__(self):
+        return len(self.struct.fields)
+
+    def __str__(self):
+        schema_str = ""
+        for field in self.fields:
+            schema_str += f"{field.field_id}: name={field.name}, type={field.type}, required={field.is_required}\n"
+        return schema_str.rstrip()
+
+    def __repr__(self):
+        return f"Schema(fields={repr(self.fields)}, schema_id={self.schema_id})"
+
+    @property
+    def fields(self) -> Tuple:
+        return self._struct.fields
```

Comment:
Why does this deviate from the Java API? Renaming `columns` to `fields` doesn't seem like a necessary change to me. I would opt to use mostly the same names for methods where possible so that it is easy to go between the two APIs.